source/common/config/grpc_mux_impl.cc

```diff
@@ -116,17 +116,26 @@ ScopedResume GrpcMuxImpl::pause(const std::vector<std::string> type_urls) {
 void GrpcMuxImpl::onDiscoveryResponse(
     std::unique_ptr<envoy::service::discovery::v3::DiscoveryResponse>&& message,
     ControlPlaneStats& control_plane_stats) {
-  const std::string& type_url = message->type_url();
+  std::string type_url = message->type_url();
   ENVOY_LOG(debug, "Received gRPC message for {} at version {}", type_url, message->version_info());
   if (message->has_control_plane()) {
     control_plane_stats.identifier_.set(message->control_plane().identifier());
   }
-  if (api_state_.count(type_url) == 0) {
-    ENVOY_LOG(warn, "Ignoring the message for type URL {} as it has no current subscribers.",
-              type_url);
-    // TODO(yuval-k): This should never happen. consider dropping the stream as this is a
-    // protocol violation
-    return;
+  // If this type url is not watched(no subscriber or no watcher), try older version of type url.
+  if (api_state_.count(type_url) == 0 ||
+      (api_state_[type_url].watches_.empty() && !message->resources().empty())) {
+    absl::optional<std::string> old_type_url = ApiTypeOracle::getEarlierTypeUrl(type_url);
+    if (old_type_url.has_value() && api_state_.count(*old_type_url)) {
+      ENVOY_LOG(debug, "v3 {} converted to v2 {}.", message->type_url(), old_type_url.value());
+      type_url = old_type_url.value();
+    }
+    if (api_state_.count(type_url) == 0) {
```

Comment:
So, thinking back to our conversation the other day, we definitely don't need to worry about v4alpha, so v2/v3 is fine as you have it. But, one thing that might not be treated correctly here is that we could have an EDS subscription to v3, but a v2 type URL appears in the message. Does this need special treatment?

Reply:
The issue doesn't cover the case where we have v2 response to v3 request, this can be fixed easily though.

Reply:
But I guess we currently don't have any interface to convert a v2 type url to a v3 type url?

Reply:
We can always downgrade type url to v2 in addWatch(), which can solve this problem, but it will cause performance penalty for common use cases. For example, v2 response to v2 request, v3 response to v3 request.